src/HTTPURLEncodedBodyParser.cpp

```diff
@@ -0,0 +1,125 @@
+#include "HTTPURLEncodedBodyParser.hpp"
+
+#define CHUNKSIZE 512
+#define MINCHUNKSIZE 64
+
+namespace httpsserver {
+
+HTTPURLEncodedBodyParser::HTTPURLEncodedBodyParser(HTTPRequest * req):
+  HTTPBodyParser(req),
+  bodyBuffer(NULL),
+  bodyPtr(NULL),
+  bodyLength(0),
+  fieldBuffer(""),
+  fieldPtr(NULL),
+  fieldRemainingLength(0)
+{
+  bodyLength = _request->getContentLength();
+  if (bodyLength) {
+    // We know the body length. We try to read that much and give an error if it fails.
+    bodyBuffer = (char *)malloc(bodyLength+1);
+    if (bodyBuffer == NULL) {
+      HTTPS_LOGE("HTTPURLEncodedBodyParser: out of memory");
+      return;
+    }
+    bodyPtr = bodyBuffer;
+    size_t toRead = bodyLength;
+    while(toRead > 0) {
+      size_t didRead = _request->readChars(bodyPtr, toRead);
+      if (didRead == 0) {
+        HTTPS_LOGE("HTTPURLEncodedBodyParser: short read");
+        bodyLength = bodyPtr - bodyBuffer;
+        break;
+      }
+      bodyPtr += didRead;
+      toRead -= didRead;
+    }
+  } else {
+    // We don't know the length. Read as much as possible.
+    bodyBuffer = (char *)malloc(CHUNKSIZE+1);
+    if (bodyBuffer == NULL) {
+      HTTPS_LOGE("HTTPURLEncodedBodyParser: out of memory");
+      return;
+    }
+    bodyPtr = bodyBuffer;
+    size_t bufferUsed = 0;
+    size_t bufferAvailable = CHUNKSIZE;
+    while(!_request->requestComplete()) {
+      if (bufferAvailable < MINCHUNKSIZE) {
+        bodyBuffer = (char *)realloc(bodyBuffer, bufferUsed + CHUNKSIZE+1);
+        if (bodyBuffer == NULL) {
+          HTTPS_LOGE("HTTPURLEncodedBodyParser: out of memory");
+          return;
+        }
```

Comment:
Potential memory leak. If `realloc` returns `NULL`, the `bodyBuffer` isn't `free`d but just left unmodified, so you will lose your reference to the memory area.

```suggestion
        char *pBuf = (char *)realloc(bodyBuffer, bufferUsed + CHUNKSIZE+1);
        if (pBuf == NULL) {
          HTTPS_LOGE("HTTPURLEncodedBodyParser: out of memory");
          free(bodyBuffer);
          bodyBuffer = NULL;
          return;
        }
        bodyBuffer = pBuf;
```

Reply:
Wow.... I never knew that, I've been using that pattern for ever....